File: Python/kojun.py

def solve_kojun(board, regions):
    rows, cols = len(board), len(board[0])
    region_cells = {}

    # Pré-processamento: mapeia cada região para suas células
    for r in range(rows):
        for c in range(cols):
            region = regions[r][c]
            if region not in region_cells:
                region_cells[region] = []
            region_cells[region].append((r, c))

    def backtrack():
        find = find_empty(board)
        if not find:
            return True

        row, col = find
        region = regions[row][col]
        max_val = len(region_cells[region])  # Máximo valor permitido na região

        for val in range(1, max_val + 1):
            if valid(board, regions, (row, col), val, region_cells):
                board[row][col] = val
                if backtrack():
                    return True
                board[row][col] = 0  # desfaz (backtrack)

        return False

    if backtrack():
        return board
    else:
        print("No solution exists")

def valid(board, regions, pos, val, region_cells):
    row, col = pos
    region = regions[row][col]

    # Regra 1: único na região
    for (rr, cc) in region_cells[region]:
        if board[rr][cc] == val:
            return False

    # Regra 2: adjacentes diferentes
    for dr, dc in [(-1,0), (1,0), (0,-1), (0,1)]:
        nr, nc = row + dr, col + dc
        if 0 <= nr < len(board) and 0 <= nc < len(board[0]) and board[nr][nc] == val:
            return False

    # Regra 3: ordem vertical
    # Célula acima
    if row > 0 and regions[row][col] == regions[row - 1][col]:
        up_val = board[row - 1][col]
        if up_val != 0 and val != 0 and up_val <= val:
            return False
    # Célula abaixo
    if row < len(board) - 1 and regions[row][col] == regions[row + 1][col]:
        down_val = board[row + 1][col]
        if down_val != 0 and val != 0 and val <= down_val:
            return False

    return True


def find_empty(board):
    for i in range(len(board)):
        for j in range(len(board[i])):
            if board[i][j] == 0:
                return (i, j)  # linha, coluna
    return None

File: Haskell/Kojun.hs
-- Módulo Kojun.hs: Implementa a lógica de resolução do puzzle Kojun
module Kojun (solveKojun, Grid, RegionMap) where

type Coord = (Int, Int)
type Grid = [[Int]]
type RegionMap = [[Int]]

-- Função principal: resolve o puzzle Kojun
solveKojun :: Grid -> RegionMap -> Maybe Grid
solveKojun grid regions = backtrack grid regions (0, 0)


backtrack :: Grid -> RegionMap -> Coord -> Maybe Grid
backtrack grid regions (r, c)
  -- caso base: se já percorreu todas as linhas, retorna o grid
  | r == length grid = Just grid
  -- se a célula grid[r][c] já estiver preenchida, continua para a próxima
  | grid !! r !! c /= 0 = backtrack grid regions (nextCoord grid (r, c))
  | otherwise =
      -- celula vazia, pegamos o id da regiao
      let regionId = regions !! r !! c
          -- calcula o tamanho da região
          regionSize = length [ () | i <- [0..length grid - 1], j <- [0..length grid - 1], regions !! i !! j == regionId ]
          -- gera os candidatos possíveis (quantidade de números na região)
          candidates = [1..regionSize]
          -- para cada candidato, atualiza o grid e tenta resolver, colocando cada v candidato na posiçao
          -- e gerando possibilidades de grids
          nextGrids = [ updateGrid grid (r, c) v | v <- candidates, valid grid regions (r, c) v ]
      -- passo recursivo
      in trySolutions nextGrids regions (nextCoord grid (r, c))


-- Tenta soluções possíveis
trySolutions :: [Grid] -> RegionMap -> Coord -> Maybe Grid
-- caso base: a lista de grids esta vazia, retorna Nothing
trySolutions [] _ _ = Nothing
-- passo recursivo: testa o primeiro grid da lista
trySolutions (g:gs) regions coord =
  -- chama backtrak recursivamente para o grid atual, a mesma regiao e a proxima coordenada
  case backtrack g regions coord of
    -- se o backtrack com esse grid deu certo, retornamos, interrompendo a busca,
    -- e não testamos as demais tentativas(gs), pois uma ja funcionou!
    Just solution -> Just solution
    -- se não, continuamos testando os demais grids
    Nothing -> trySolutions gs regions coord


-- Calcula a próxima coordenada no grid
nextCoord :: Grid -> Coord -> Coord
nextCoord grid (r, c)
  -- se já percorreu todas as colunas, vai para a próxima linha
  | c < length (head grid) - 1 = (r, c + 1)
  | otherwise                  = (r + 1, 0)


-- Atualiza uma célula do grid com um novo valor
updateGrid :: Grid -> Coord -> Int -> Grid
updateGrid grid (r, c) val =
  -- pega todas as linhas acima da linha r (exclusiva)
  take r grid ++
  -- pega a linha r, substitui o valor na coluna c e concatena com o restante da linha
  [take c (grid !! r) ++ [val] ++ drop (c + 1) (grid !! r)] ++
  -- pega todas as linhas abaixo da linha r
  drop (r + 1) grid
  -- Na concatenação, retorna o grid com o novo valor na posição (r, c)


-- Verifica se o valor é válido na posição dada
valid :: Grid -> RegionMap -> Coord -> Int -> Bool
valid grid regions coord@(r, c) val =
  -- testa as 3 regras do puzzle Kojun
  notSameAsAdjacent && notInRegion && validVertical grid regions coord val
  where
    notSameAsAdjacent =
      -- verifica se o valor não é igual aos vizinhos
      getCell grid (r - 1, c) /= Just val &&
      getCell grid (r + 1, c) /= Just val &&
      getCell grid (r, c - 1) /= Just val &&
      getCell grid (r, c + 1) /= Just val

    regionId = regions !! r !! c
    -- gera uma lista com todas as coordenadas da regiao atual de (r, c)
    regionCells = [ (i, j) | i <- [0..length grid - 1], j <- [0..length grid - 1], regions !! i !! j == regionId ]
    -- filtra os valores já preenchidas na região
    regionVals = [ grid !! i !! j | (i, j) <- regionCells, grid !! i !! j /= 0 ]
    -- verifica se o valor não está presente na região
    notInRegion = val `notElem` regionVals


-- Validação da ordem vertical dentro da mesma região
validVertical :: Grid -> RegionMap -> Coord -> Int -> Bool
validVertical grid regions (r, c) val =
  let upRegion    = getRegion regions (r - 1, c)
      downRegion  = getRegion regions (r + 1, c)
  -- primeiro analisamos a celula acima
  in case getCell grid (r - 1, c) of
      -- se a celula contem um valor u, então, u não pode ser igual a val
      -- e se a região acima for diferente da região atual, u deve ser maior que val
      Just u -> u /= val && (upRegion /= Just regionId || u > val)
      -- se a celula acima não existe (fora dos limites ou fora da região), não há restricao
      Nothing -> True
    -- continua com a validaçao, testando a regra da celula abaixo no mesmo in
    &&
    case getCell grid (r + 1, c) of
      -- se a celula contem um valor d, então, d não pode ser igual a val
      -- e se a região abaixo for diferente da região atual, d deve ser menor que val
      Just d -> d /= val && (downRegion /= Just regionId || val > d)
      -- novamente, se a celula acima não existe (fora dos limites ou fora da região), não há restricao
      Nothing -> True
  where
    -- area de variavel local, calcula o ID da regiao
    regionId = regions !! r !! c


-- Acesso seguro ao valor do grid
getCell :: Grid -> Coord -> Maybe Int
getCell grid (r, c)
  | r < 0 || r >= length grid = Nothing
  | c < 0 || c >= length (head grid) = Nothing
  | otherwise = Just ((grid !! r) !! c)

-- Acesso seguro à região
getRegion :: RegionMap -> Coord -> Maybe Int
getRegion regions (r, c)
  | r < 0 || r >= length regions = Nothing
  | c < 0 || c >= length (head regions) = Nothing
  | otherwise = Just ((regions !! r) !! c)